libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp

```diff
@@ -40,169 +58,425 @@ namespace mpi = pika::mpi::experimental;
-        std::cout << temp.str() << std::endl;
+        using namespace pika::debug::detail;
+        int other = (mtype == msg_type::send) ? next_rank(rank, size) : prev_rank(rank, size);
+        const char* msg = (mtype == msg_type::send) ? "send" : "recv";
+        std::stringstream temp;
+        temp << dec<3>(rank) << "/" << dec<3>(size);
+        // clang-format off
+        msr_deb<1>.debug(str<>(temp.str().c_str())
+                         , "token", hex<4>(token)
+                         , "<-/->", dec<3>(other)
+                         , "tag", dec<3>(tag)
+                         , "round", dec<3>(round)
+                         , "step", dec<3>(step)
+                         , msg
+                         , (xmsg == nullptr) ? "" : xmsg
+                         , "counter", std::int64_t(counter));
+        // clang-format on
     }
 }
 
-void msg_send(int rank, int size, int to, int /*from*/, int token, unsigned tag)
+// ------------------------------------------------------------
+// message buffers get reused from a stack
+boost::lockfree::stack<message_buffer*, boost::lockfree::fixed_sized<false>> message_buffers(1024);
+std::atomic<std::uint64_t> message_buffers_size_{0};
+
+message_buffer* get_msg_buffer(std::uint64_t size)
 {
-    if (output)
+    message_buffer* buffer;
+    if (!message_buffers.pop(buffer))
     {
-        std::ostringstream temp;
-        temp << "Rank " << std::setfill(' ') << std::setw(3) << rank << " of "
-             << std::setfill(' ') << std::setw(3) << size << " Sent token "
-             << std::setfill(' ') << std::setw(3) << token << " to   rank "
-             << std::setfill(' ') << std::setw(3) << to << " tag "
-             << std::setfill(' ') << std::setw(3) << tag;
-        std::cout << temp.str() << std::endl;
+        // allocate the amount of space we want
+        void* data = new unsigned char[size];
+        // construct out buffer object in that space
+        buffer = new (data) message_buffer(size);
+    }
+    // set initial token to some easy to spot default
+    buffer->token_val_ = 0xc0de;
+    message_buffers_size_++;
+    msr_deb<6>.debug(str<>("message_buffers"), std::uint64_t(message_buffers_size_.load()));
+    return buffer;
+}
+
+void release_msg_buffer(message_buffer* buffer)
+{
+    --message_buffers_size_;
+    message_buffers.push(buffer);
+    msr_deb<6>.debug(str<>("message_buffers"), std::uint64_t(message_buffers_size_.load()));
+}
+
+namespace pika::execution {
+    template <typename Sender>
+    auto async(Sender&& sender)
+    {
+        namespace ex = pika::execution::experimental;
+        auto sched = ex::thread_pool_scheduler{&pika::resource::get_thread_pool("default")};
+        auto snd = ex::schedule(sched) |
+            ex::then([sender = std::move(sender)]() mutable { ex::start_detached(std::move(sender)); });
+        ex::start_detached(std::move(snd));
+    }
+}    // namespace pika::execution
```

Comment:
Same comment here as in the other test. If you really want a helper like this, call it maybe `on`, give it a scheduler as the first parameter, and just put it in the global namespace for this test. Also:
```suggestion
namespace pika::execution {
    template <typename Sender>
    auto async(Sender&& sender)
    {
        namespace ex = pika::execution::experimental;
        auto sched = ex::thread_pool_scheduler{&pika::resource::get_thread_pool("default")};
        auto snd = ex::schedule(sched) |
            ex::then([sender = std::forward<Sender>(sender)]() mutable { ex::start_detached(std::move(sender)); });
        ex::start_detached(std::move(snd));
    }
}    // namespace pika::execution
```